File: coding/src/com/algos/li05_binarysearch/revised/KThElementOfTwoSortedArrays.java
package  com.algos.li05_binarysearch.revised;

public class KThElementOfTwoSortedArrays {
    public static void main(String args[]) {
        KThElementOfTwoSortedArrays obj = new KThElementOfTwoSortedArrays();

        int nums1[] = { 2, 3, 6, 7, 9 };
        int nums2[] = { 1, 4, 8, 10 };
        int m = nums1.length;
        int n = nums2.length;
        int k = 5;
        System.out.println(obj.kthelement(nums1, nums2, m, n, k));
    }

    public int kthelement(int[] nums1, int[] nums2, int n1, int n2, int k) {
        if (n1 > n2) 
            return kthelement(nums2, nums1, n2, n1, k);

        int low = Math.max(0, k - n2), high = Math.min(k, n1);
        while (low <= high) {
            int mid1 = (low + high) >> 1;
            int mid2 = k - mid1;                // not mid-cut1 but k-cut1
            int l1 = mid1 == 0 ? Integer.MIN_VALUE : nums1[mid1 - 1];
            int l2 = mid2 == 0 ? Integer.MIN_VALUE : nums2[mid2 - 1];
            int r1 = mid1 == n2 ? Integer.MAX_VALUE : nums1[mid1];
            int r2 = mid2 == n1 ? Integer.MAX_VALUE : nums2[mid2];

            if (l1 <= r2 && l2 <= r1) {
                return Math.max(l1, l2);
            } else if (l1 > r2) {
                high = mid1 - 1;
            } else {
                low = mid1 + 1;
            }
        }
        return -1;
    }
}

File: dsa/patterns/05_binarysearch/python/kth_element_of_two_sorted_arrays.py
"""
K-th Element of Two Sorted Arrays

Idea:
Binary search on the smaller array.
Partition arrays such that:
- left part has k elements
- max(left) <= min(right)

Time Complexity: O(log(min(n, m)))
Space Complexity: O(1)
"""

import math


def kth_element(nums1, nums2, k):
    n1, n2 = len(nums1), len(nums2)

    # Always binary search on the smaller array
    if n1 > n2:
        return kth_element(nums2, nums1, k)

    low = max(0, k - n2)
    high = min(k, n1)

    while low <= high:
        cut1 = (low + high) // 2
        cut2 = k - cut1

        l1 = -math.inf if cut1 == 0 else nums1[cut1 - 1]
        l2 = -math.inf if cut2 == 0 else nums2[cut2 - 1]

        r1 = math.inf if cut1 == n1 else nums1[cut1]
        r2 = math.inf if cut2 == n2 else nums2[cut2]

        if l1 <= r2 and l2 <= r1:
            return max(l1, l2)
        elif l1 > r2:
            high = cut1 - 1
        else:
            low = cut1 + 1

    return -1


# Driver code
if __name__ == "__main__":
    nums1 = [2, 3, 6, 7, 9]
    nums2 = [1, 4, 8, 10]
    k = 5
    print(kth_element(nums1, nums2, k))
